rfcs/0024-python-scope-package-set.md

```diff
@@ -0,0 +1,145 @@
+maintainers of the package sets and the individual expressions. Another
+important factor is whether *upstream* provides a curated set of packages that
+are known to be compatible. An example of a curated set or snapshot is Stackage
+for Haskell. If such set exists and serves the needs from the community, then
+the scope may be limited to providing that set.
+
+## Definitions
+
+The following definitions are used in the scope of this RFC:
+- application: a package providing programs that happen to be written in Python
+- library: a package providing modules and programs that are used in a Python environment for development purposes or as dependency for applications
+- package set: a curated set of compatible libraries that can be used in a development environment and as dependencies for applications
+- maintainer: a maintainer of an expression as typically noted in the `meta.maintainers` field
+- package set maintainer: a maintainer of the Python package set as noted in the `CODEOWNERS` file
+
+Examples on the usage of application and library:
+- `scipy` is a collection of routines for scientific computing. It is used for development purposes or as a library and is therefore considered a *library*
+- `calibre` is a program for managing e-books that happens to be written in Python. The package is therefore considered an *application*
+- `pytest` is a test runner. While a program, it can only be used in conjunction with the Python environment where it needs to import the modules that are to be tested. It is therefore considered a *library*
+
+## Scope of the package set
+
+The purpose of the package set is to provide a set of compatible libraries that
+can be used for development purposes and as dependencies by Python applications.
+Applications are not considered part of the package set and their expressions
+shall thus also be placed elsewhere.
+
+For each NixOS release, a set of compatible libraries shall be formed
+that functions on all supported platforms, interpreter versions and
+implementations.
+
+That in effect means only one version of a package is permitted as having
+multiple versions in the set may lead to collisions. Exceptions can be made,
+like in the case of build or test-time dependencies such as the `pytest` test runner.
+
+The definitions from the previous section shall be used to decide whether an
+expression should be part of the Python package set or not. The following
+special cases exist.
+
+### Package that is used as a both an application and a library
+In certain cases a package is used as both an application and a library. The
+predominant use case determines whether the expression is considered an
+application or a library. If both use cases are considered common, then the
+expression shall be part of the package set and a special alias shall we added
+to the top-level attribute set. This special attribute will mark the package as an
+application and shall be documented in the Nixpkgs manual.
```

Comment:
I need better wording for that sentence. Basically, what I would like to say is that there shall be a function `toPythonApplication` that is documented in the Nixpkgs manual that is to be used for the top-level attribute.

E.g.
```
calibre = with python.pkgs; toPythonApplication calibre;
```

Reply:
What this function will actually do? Create `.desktop` entries? What about pure-CLI tools?

And what you meant is «There shall be a function to perform additional post-installation steps specific to applications written in Python. This function shall be documented in the Nixpkgs manual.»?

Reply:
Thanks. I've rewritten that section. Basically, it is important to distinguish applications from libraries because, when a library as part of a development environment needs to call an application, we don't want the dependencies of the application to be added to the Python environment. Or if one application *written* in Python calls another *written* in Python.

As this is an implementation detail (the use of `propagatedBuildInputs`) I on purpose remain vague in this description.

Reply:
I think that the need for propagation of dependencies is unlikely to change in Python2 or Python3, so even if it is an implementation detail, it could be mentioned somewhere, both in the manual (eventually) and in the RFC.

Reply:
Actually, there is the `requiredPythonModules` function in `python-packages.nix` which is also already used by `python.buildEnv` and could also be used in `buildPython*`. That could drop the need for `propagatedBuildInputs` although it likely increases the evaluation time.